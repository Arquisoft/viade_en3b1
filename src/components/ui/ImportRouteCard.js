import React from 'react'
import { Card, CardContent, IconButton, CardMedia, Typography, makeStyles, CardActions, Collapse, Box } from '@material-ui/core';
import MapSnapshot from '../map/MapSnapshot';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import clsx from 'clsx';

function ImportRouteCard(props) {
    const classes = useStyles();
    const { route } = props;

    const [expanded, setExpanded] = React.useState(false);

    const handleExpandClick = () => {
        setExpanded(!expanded);
    };

    return (
        <div>
            <Card elevation={4} className={classes.root}>
                <div className={classes.details}>
                    <CardContent className={classes.content}>
                        <Typography className={classes.title} component="h5" variant="h5">
                            {route.getName()}
                        </Typography>
                    </CardContent>

                    <CardActions disableSpacing>
                        <IconButton
                            data-testid="btn-expand-test"
                            className={clsx(classes.expand, {
                                [classes.expandOpen]: expanded,
                            })}
                            onClick={handleExpandClick}
                            aria-expanded={expanded}
                            aria-label="show more"
                        >
                            <ExpandMoreIcon />
                        </IconButton>
                    </CardActions>
                    <Collapse in={expanded} timeout="auto" unmountOnExit>
                        <CardContent >
                            {(route.getDescription()) ? (
                                <div>
                                    <Box fontSize="h6.fontSize" fontWeight="fontWeightBold">
                                        Description:
                                    </Box>
                                    <Box fontWeight="fontWeightRegular">
                                        {route.getDescription()}
                                    </Box>
                                </div>
                            ) : (
                                    <Box fontWeight="fontWeightRegular">
                                        No description provided.
                                    </Box>
                                )}
                        </CardContent>
                    </Collapse>
                </div>
                <CardMedia
                    className={classes.cover}
                >
                    <MapSnapshot route={route}></MapSnapshot>
                </CardMedia>
            </Card>
        </div>
    )
}

const useStyles = makeStyles((theme) => ({
    root: {
        display: 'flex',
        marginTop: theme.spacing(5)
    },
    details: {
        display: 'flex',
        flexDirection: 'column',
        width: theme.spacing(50)
    },
    content: {
        flex: '1 0 auto',
    },
    title: {
        martinTop: theme.spacing(5)
    },
    cover: {
        width: theme.spacing(30),
        marinLeft: 'auto',
    },
    expand: {
        transform: 'rotate(0deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    expandOpen: {
        transform: 'rotate(180deg)',
    },
}));

export default ImportRouteCard;