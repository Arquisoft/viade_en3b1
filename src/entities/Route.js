import { v4 as uuid } from "uuid";
import RouteElement from "./RouteElement";

class Route {

    /**
     * Constructor for a new Route. 
     * An id will be automatically generated (uuid v4).
     * TotalDistance will also be automatically generated.
     * 
     * RouteElements have name, latitude, longitude and elevation.
     * 
     * @param {String} name Name of the route.
     * @param {Date} date Date of the route.
     * @param {String} description Description of the route.
     * @param {Array} routeElements List of RouteElement objects, which are the trackpoints.
     * @param {Array} comments List of comments
     * @param {Array<String>} media List of urls of media (on the POD)
     */
    constructor(name, description, routeElements, comments, media, date) {
        this.id = uuid().toString();
        this.name = name;
        this.date = date;
        this.description = description;

        this.media = media;
        this.totalDistance = null;

        if (comments === null) {
            this.comments = [];
        } else {
            this.comments = comments;
        }

        if (routeElements[0] instanceof RouteElement) {
            this.routeElements = routeElements;
        } else {
            this.routeElements = generateRouteElements(routeElements);
        }
    }

    getId() {
        return this.id;
    }

    getName() {
        return this.name;
    }

    getDate() {
        return this.date;
    }

    getTime() {
        return this.time;
    }

    getDescription() {
        return this.description;
    }

    getRouteElements() {
        return this.routeElements;
    }

    getTotalDistance() {
        var total = 0;
        var aux = 0;

        for (let i = 1; i < this.routeElements.length; i++) {
            aux = this.routeElements[i].getLatitude() - this.routeElements[i - 1].getLatitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getLongitude() - this.routeElements[i - 1].getLongitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getElevation() - this.routeElements[i - 1].getElevation();
            total += this.addDistance(aux);
        }

        return total;
    }

    addDistance(aux) {
        if (aux < 0) {
            return -aux;
        } else {
            return aux;
        }
    }

    getComments() {
        return this.comments;
    }

    getMedia() {
        return this.media;
    }

    addMedia(media) {
        this.media.push(media);
    }

    addRouteElement(routeElement) {
        this.routeElements.push(routeElement);
    }

    getJsonLD() {
        let routePointsJson = [];
        this.routeElements.forEach((p) => routePointsJson.push(p.toJsonLatLng()));
        return JSON.stringify(
            {
                "@context": {
                    "@version": 1.1,
                    "comments": {
                        "@container": "@list",
                        "@id": "viade:comments"
                    },
                    "description": {
                        "@id": "schema:description",
                        "@type": "xs:string"
                    },
                    "media": {
                        "@container": "@list",
                        "@id": "viade:media"
                    },
                    "name": {
                        "@id": "schema:name",
                        "@type": "xs:string"
                    },
                    "points": {
                        "@container": "@list",
                        "@id": "viade:points"
                    },
                    "latitude": {
                        "@id": "schema:latitude",
                        "@type": "xs:double"
                    },
                    "longitude": {
                        "@id": "schema:longitude",
                        "@type": "xs:double"
                    },
                    "elevation": {
                        "@id": "schema:elevation",
                        "@type": "xsd:double"
                    },
                    "author": {
                        "@id": "schema:author",
                        "@type": "@id"
                    },
                    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                    "schema": "http://schema.org/",
                    "viade": "http://arquisoft.github.io/viadeSpec/",
                    "xsd": "http://www.w3.org/2001/XMLSchema#"
                },
                "name": this.name,
                "author": this.author,
                "description": this.description,
                "comments": this.comments,
                "media": this.media,
                "points": routePointsJson
            }
        );
    }

}

/**
 * Method that receives a list of points and returns
 * a list of RouteElements
 * @param {Array} points List of points containing name, latitude, longitude, elevation. 
 */
function generateRouteElements(points) {
    let routeElements = [];
    points.forEach((p) => routeElements.push(new RouteElement(p.lat, p.lng)));
    return routeElements;
}

export default Route;