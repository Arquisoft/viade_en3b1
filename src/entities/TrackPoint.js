class TrackPoint {

    /**
     * 
     * @param {Number} latitude 
     * @param {Number} longitude 
     * @param {String} name 
     * @param {Number} elevation 
     */
    constructor(latitude, longitude, name, elevation) {

        this.latitude = latitude;
        this.longitude = longitude;

        if (!name) {
            this.name = "";
        } else {
            this.name = name;
        }

        if(!elevation){
            this.elevation = "";
        } else {
            this.elevation = elevation;
        }
    }

    getName() {
        return this.name;
    }

    getLatitude() {
        return this.latitude;
    }

    getElevation() {
        return this.elevation;
    }

    getLongitude() {
        return this.longitude;
    }

    setElevation(num) {
        this.elevation = num;
    }

    toJson() {
        return {
            "latitude": this.latitude,
            "longitude": this.longitude,
            "elevation": this.elevation,
        };
    }
}

export default TrackPoint;