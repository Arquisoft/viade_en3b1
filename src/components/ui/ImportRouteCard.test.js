import ImportRouteCard from './ImportRouteCard';
import Route from '../../entities/Route';
import TrackPoint from '../../entities/TrackPoint';
import { render, fireEvent, getByTestId } from '@testing-library/react';

var route1 = new Route("Route1", "Description 1", [new TrackPoint(1,1), new TrackPoint(2,1)], null, null, null, "ID1");

test('Define ImportRouteCard', () => {
  expect(ImportRouteCard).toBeDefined();
});

test('Renders correctly', () => {
  const wrapper = render(<ImportRouteCard route={route1} />);
  expect(wrapper).toMatchSnapshot();
});

test('Click on expand', () => {
    const { container } = render(<ImportRouteCard route={route1}/>);

    const button = getByTestId(container, 'btn-expand-test');
    fireEvent.click(button);
});