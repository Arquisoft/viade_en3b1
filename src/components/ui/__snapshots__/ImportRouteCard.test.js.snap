// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation4 MuiPaper-rounded"
        >
          <div
            class="makeStyles-details-2"
          >
            <div
              class="MuiCardContent-root makeStyles-content-3"
            >
              <h5
                class="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
              >
                Route1
              </h5>
            </div>
            <div
              class="MuiCardActions-root"
            >
              <button
                aria-expanded="false"
                aria-label="show more"
                class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-6"
                data-testid="btn-expand-test"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="MuiCardMedia-root makeStyles-cover-5"
          >
            <div
              class="leaflet-container"
              style="position: relative;"
            >
              <div
                class="leaflet-pane leaflet-map-pane"
                style="left: 0px; top: 0px;"
              >
                <div
                  class="leaflet-pane leaflet-tile-pane"
                >
                  <div
                    class="leaflet-layer "
                    style="z-index: 1;"
                  >
                    <div
                      class="leaflet-tile-container leaflet-zoom-animated"
                      style="z-index: 18; left: 0px; top: 0px;"
                    >
                      <img
                        alt=""
                        class="leaflet-tile"
                        role="presentation"
                        src="https://c.tile.openstreetmap.org/18/131800/129979.png"
                        style="width: 256px; height: 256px; left: -46px; top: -156px;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="leaflet-pane leaflet-shadow-pane"
                />
                <div
                  class="leaflet-pane leaflet-overlay-pane"
                >
                  <svg
                    height="0"
                    pointer-events="none"
                    style="left: 0px; top: 0px;"
                    viewBox="0 0 0 0"
                    width="0"
                  >
                    <g>
                      <path
                        class="leaflet-interactive"
                        d="M0 0L0 0"
                        fill="none"
                        stroke="blue"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-opacity="1"
                        stroke-width="3"
                      />
                    </g>
                  </svg>
                </div>
                <div
                  class="leaflet-pane leaflet-marker-pane"
                />
                <div
                  class="leaflet-pane leaflet-tooltip-pane"
                />
                <div
                  class="leaflet-pane leaflet-popup-pane"
                />
              </div>
              <div
                class="leaflet-control-container"
              >
                <div
                  class="leaflet-top leaflet-left"
                />
                <div
                  class="leaflet-top leaflet-right"
                />
                <div
                  class="leaflet-bottom leaflet-left"
                />
                <div
                  class="leaflet-bottom leaflet-right"
                >
                  <div
                    class="leaflet-control-attribution leaflet-control"
                  >
                    <a
                      href="https://leafletjs.com"
                      title="A JS library for interactive maps"
                    >
                      Leaflet
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation4 MuiPaper-rounded"
      >
        <div
          class="makeStyles-details-2"
        >
          <div
            class="MuiCardContent-root makeStyles-content-3"
          >
            <h5
              class="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
            >
              Route1
            </h5>
          </div>
          <div
            class="MuiCardActions-root"
          >
            <button
              aria-expanded="false"
              aria-label="show more"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-6"
              data-testid="btn-expand-test"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiCardMedia-root makeStyles-cover-5"
        >
          <div
            class="leaflet-container"
            style="position: relative;"
          >
            <div
              class="leaflet-pane leaflet-map-pane"
              style="left: 0px; top: 0px;"
            >
              <div
                class="leaflet-pane leaflet-tile-pane"
              >
                <div
                  class="leaflet-layer "
                  style="z-index: 1;"
                >
                  <div
                    class="leaflet-tile-container leaflet-zoom-animated"
                    style="z-index: 18; left: 0px; top: 0px;"
                  >
                    <img
                      alt=""
                      class="leaflet-tile"
                      role="presentation"
                      src="https://c.tile.openstreetmap.org/18/131800/129979.png"
                      style="width: 256px; height: 256px; left: -46px; top: -156px;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="leaflet-pane leaflet-shadow-pane"
              />
              <div
                class="leaflet-pane leaflet-overlay-pane"
              >
                <svg
                  height="0"
                  pointer-events="none"
                  style="left: 0px; top: 0px;"
                  viewBox="0 0 0 0"
                  width="0"
                >
                  <g>
                    <path
                      class="leaflet-interactive"
                      d="M0 0L0 0"
                      fill="none"
                      stroke="blue"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-opacity="1"
                      stroke-width="3"
                    />
                  </g>
                </svg>
              </div>
              <div
                class="leaflet-pane leaflet-marker-pane"
              />
              <div
                class="leaflet-pane leaflet-tooltip-pane"
              />
              <div
                class="leaflet-pane leaflet-popup-pane"
              />
            </div>
            <div
              class="leaflet-control-container"
            >
              <div
                class="leaflet-top leaflet-left"
              />
              <div
                class="leaflet-top leaflet-right"
              />
              <div
                class="leaflet-bottom leaflet-left"
              />
              <div
                class="leaflet-bottom leaflet-right"
              >
                <div
                  class="leaflet-control-attribution leaflet-control"
                >
                  <a
                    href="https://leafletjs.com"
                    title="A JS library for interactive maps"
                  >
                    Leaflet
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
